### PR TITLE
Throw when compiling spread in `super()` but not classes

### LINKED ISSUE #4254: Should work when a child class has class properties and no constructor, and transpiling spreads on super. (T7309)
> Issue originally made by Paul Sanchez (basicdays)
### Bug information
- **Babel version:** 6.7.7
- **Node version:** 4.4.2
- **npm version:** 3.8.6
### Options

```
babel-plugin-transform-class-properties
babel-plugin-transform-es2015-spread
```
### Input code

```
class Parent {}

class Child extends Parent {
    prop = 'val';
}
```
### Description

When using class properties on a child class without an explicit constructor, a default constructor is created by Babel that includes a spread in the call in `super`. This causes a syntax error in Node 4. Here's the transpiled output:

```
'use strict';

class Parent {}

class Child extends Parent {
        constructor() {
                var _temp;

                return _temp = super(...arguments), this.prop = 'val', _temp;
        }

}
```

A temporary workaround is to also include `babel-plugin-transform-es2015-classes` or to explicitly provide `constructor() { super() }` in the child class.

### LINKED ISSUE #4375: transform-es2015-spread does not transform super call (T7467)
> Issue originally made by @Kovensky
### Bug information
- **Babel version:** 6.10.4
- **Node version:** 4.4.5
- **npm version:** 2.15.4
### Options

```
--plugins transform-es2015-parameters,transform-es2015-spread
```
### Input code

```
class Child extends Parent {
  constructor (...args) {
    super(...args)
    notSuper(...args)
  }
}
```
### Description

`transform-es2015-spread` ignores, or otherwise is unable to, transform spreads to a `super` call.

This is related to https://phabricator.babeljs.io/T7309 as the default constructor generated by `transform-class-properties` starts from  `constructor (...args) { super(...args) }`, but it happens even if you write that particular constructor without that plugin being involved.

Adding `transform-es2015-classes` converts the `super` call into (something similar to) `Object.setPrototypeOf(Child, Parent); Object.getPrototypeOf(Child).call(this, ...args)`, which the `transform-es2015-spread` correctly transforms.

In case it is completely impossible to transform spreads into `super`, perhaps `transform-es2015-spread` should complain if a spread `super` call is used without `transform-es2015-classes`. Right now it silently generates bad code.

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Closes #4375, closes #4254
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Babel already generates invalid code (it leaves `...` even if it's a syntax error in older browsers), so it's better to throw an explicit error explaining how to fix it.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12722"></a>

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
I'm not sure about how to label this PR

**Review by existentialism — APPROVED** [review]

File: packages/babel-plugin-transform-spread/src/index.js [hunk 1]
```diff
@@ -97,7 +97,12 @@ export default declare((api, options) => {
 
         const calleePath = skipTransparentExprWrappers(path.get("callee"));
 
-        if (calleePath.isSuper()) return;
+        if (calleePath.isSuper()) {
+          throw path.buildCodeFrameError(
+            "It's not possible to compile spread arguments in `super()` without compiling classes.\n" +
+              "Please add '@babel/plugin-transform-classes' to your Babel configuration.",
```
Comment (JLHwung) [inline packages/babel-plugin-transform-spread/src/index.js:103]: The classes and spread transform have almost identical compat data:

```json
{
"transform-classes": {
    "chrome": "46",
    "opera": "33",
    "edge": "13",
    "firefox": "45",
    "safari": "10",
    "node": "5",
    "ios": "10",
    "samsung": "5",
    "electron": "0.36"
  },
"transform-spread": {
    "chrome": "46",
    "opera": "33",
    "edge": "13",
    "firefox": "36",
    "safari": "10",
    "node": "5",
    "ios": "10",
    "samsung": "5",
    "electron": "0.36"
  },
}
```
Thus it is very unlikely that users of `@babel/preset-env` will hit this issue: if `transform-spread` is enabled, `transform-classes` is also enabled, unless explicitly disabled by `exclude`.

Leave a note here for reference only.

**Review by JLHwung — APPROVED** [review]

**Comment by wardpeet** [comment]
This breaks when you're using `@babel/plugin-proposal-class-properties`.

**Comment by phaistonian** [comment]
Hello.

Is this fix pushed in production yet?

I had to downgrade for `7.12.1` to _avoid_ work around it.

**Comment by nicolo-ribaudo** [comment]
By "this fix" do you mean "throwing an explicit error" or "support compiling `spread()` in super without compiling classes"?

**Comment by phaistonian** [comment]
The latter.

**Comment by nicolo-ribaudo** [comment]
Ok not yet, I'm sorry.

> I had to downgrade for `7.12.1` to avoid work around it.

Note that Babel 7.12.1 generates invalid code: this PR made it a clear error rather than just silently doing the wrong thing.
If you need to compile the spread operator you need to also compile classes for now.

**Pull request merged**